presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -447,15 +452,31 @@ private OuterJoinPushDownResult processLimitedOuterJoin(Expression inheritedPred
                 }
             }
 
-            // See if we can push down any outer or join predicates to the inner side
-            for (Expression conjunct : EqualityInference.nonInferrableConjuncts(and(outerEffectivePredicate, joinPredicate))) {
+            // See if we can push down any outer predicates to the inner side
+            for (Expression conjunct : EqualityInference.nonInferrableConjuncts(outerEffectivePredicate)) {
                 Expression rewritten = potentialNullSymbolInference.rewriteExpression(conjunct, not(in(outerSymbols)));
                 if (rewritten != null) {
                     innerPushdownConjuncts.add(rewritten);
                 }
             }
 
-            // TODO: consider adding join predicate optimizations to outer joins
+            // See if we can push down any join predicates to the inner side
+            for (Expression conjunct : extractConjuncts(joinPredicate)) {
+                Expression rewritten = innerInference.rewriteExpression(conjunct, not(in(outerSymbols)));
+                if (rewritten != null) {
+                    innerPushdownConjuncts.add(rewritten);
+                    if (!shouldRemoveOuterJoinConjunct(outerSymbols, conjunct)) {
```

Comment:
It would not work (well it would work - but suboptimal). Actually original code looked more or less like what you suggest.
```java
           // See if we can push down any outer or join predicates to the inner side
            for (Expression conjunct : EqualityInference.nonInferrableConjuncts(and(outerEffectivePredicate, joinPredicate))) {
                Expression rewritten = potentialNullSymbolInference.rewriteExpression(conjunct, not(in(outerSymbols)));
                if (rewritten != null) {
                    innerPushdownConjuncts.add(rewritten);
                }
            }
```

But it did not remove anything from `joinPredicate`.
There are two problems with use of `nonInferrableConjuncts`:

1) We want to pushdown to inner side and remove from joinPredicate some conjuncts which are *inferrable* in understanding of `nonInferrableConjuncts`.
E.g `b1 = b2` or `b1 = 5` or `b1 + b2 = 10` (when `b1` and `b2` are columns from inner table).
Those conjuncts would not be subject for rewrite if we used only `nonInferrableConjuncts`.

2) For other *inferrable* conjuncts we want to just do the pushdown but leave conjunct in join clause
E.g `a=b` should be pushed to inner side as `b=5` if we know from inheritedPredicate that `a=5`.
Again this expression would not be taken into consideration if we operate just on `nonInferrableConjuncts`.

The (2) could not be very important - but still beneficial.

But we definitely want (1).
So query like `SELECT .... FROM t1 JOIN t2 on t1.a = t2.b AND t2.c = t2.d` executes quickly without use of `filterFunction` evaluation during execution. This can become less relevant when generated classes for lookup join support filterFunction. But for now I would rather have this.